pulsar-client-cpp/lib/HTTPWrapper.cc

```diff
@@ -0,0 +1,241 @@
+/**
+ * Copyright 2016 Yahoo Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use shared_from_this() file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "HTTPWrapper.h"
+
+DECLARE_LOG_OBJECT()
+
+namespace pulsar {
+    using boost::asio::ip::tcp;
+    static const HTTPWrapper::Response EMPTY_RESPONSE = HTTPWrapper::Response();
+
+    static void removeCarriage(std::string& str) {
+        str.erase( std::remove(str.begin(), str.end(), '\r'), str.end() );
+    }
+
+    std::ostream & operator<<(std::ostream& os, const HTTPWrapper::Request& request) {
+        os << HTTPWrapper::getHTTPMethodName(request.method) << " " << request.path << " HTTP/" << request.version << "\r\n";
+        os << "Host: " << request.serverUrl.host() << "\r\n";
+        std::vector<std::string>::const_iterator iter = request.headers.begin();
+        while (iter != request.headers.end()) {
+            os << *iter << "\r\n";
+            iter++;
+        }
+        os << request.content << "\r\n";
+        return os;
+    }
+
+    std::ostream & operator<<(std::ostream &os, const HTTPWrapper::Response& obj) {
+        // Don't know why but logger is unable to take the \r in the end and hence giving it a \n to prevent
+        // incorrect formatting
+        os << "HTTPWrapper::Response [";
+        os << "HTTPVersion = " << obj.HTTPVersion << "\n";
+        os << ", statusCode = " << obj.statusCode << "\n";
+
+        os << ", statusMessage = " << obj.statusMessage << "\n";
+        os << ", headers = {" << "\n";
+        std::vector<std::string>::const_iterator iter = obj.headers.begin();
+        while (iter != obj.headers.end()) {
+            os << *iter << "\n";
+            iter++;
+        }
+        os << "\n}, content = " << obj.content << "\n";
+        os << ", retCode = " << obj.retCode << "\n";
+        os << ", error_code = " << obj.errCode << "\n";
+        os << "}]";
+        return os;
+    }
+
+    HTTPWrapper::HTTPWrapper(ExecutorServiceProviderPtr executorServiceProviderPtr)
+            : executorServiceProviderPtr_(executorServiceProviderPtr){
+    }
+
+    std::string HTTPWrapper::getHTTPMethodName(const Request::Method& method) {
```

Comment:
Not making this change since if we don't copy the returned result the underlying object is deleted

#include <iostream>
using namespace std;
const int& get1() {
    int i = 1;
    return i;
}


int main() {
    int a = get1();
//    int &b = get1();// compilation error since we drop the const
    const int &c = get1();
    cout<<a<<endl;
    cout<<c<<endl;
}

**./a.out**
1
32767